src/agent/Core/Controller/CheckoutSession.cpp

```diff
@@ -173,12 +174,53 @@ Controller::maybeSend100Continue(Client *client, Request *req) {
 	}
 }
 
+typedef boost::tuple<Controller*, Client*, Request*> ControllerFinishInitiatingSessionCallbackPointers;
+
+void
+Controller::onWritable(EV_P_ ev_io *io, int revents) {
+        ControllerFinishInitiatingSessionCallbackPointers *cbs = static_cast<ControllerFinishInitiatingSessionCallbackPointers *>(io->data);
+	    Controller* that = cbs->get<0>();
+		Request * req = cbs->get<2>();
+        ev_io_stop(that->getLoop(), io);
+
+		int connect_error = 0;
+		socklen_t connect_error_len = sizeof(connect_error);
+		if (-1 == getsockopt(req->session->fd(), SOL_SOCKET, SO_ERROR, &connect_error, &connect_error_len)) {
+			int err = errno;
+			throw SystemException("Cannot check socket status", err);
+		} else if (connect_error != 0) {
+			// connect_error uses the same error codes as errno
+			throw SystemException("Cannot connect socket", connect_error);
+		}
+
+		that->finishInitiatingSession(cbs->get<1>(), req);
+		delete cbs;
+}
+
+void
+Controller::finishInitiatingSession(Client *client, Request *req) {
+	TRACE_POINT();
+	SKC_DEBUG(client, "Session initiated: fd=" << req->session->fd());
+	req->appSink.reinitialize(req->session->fd());
+	req->appSource.reinitialize(req->session->fd());
+	/***************/
+	/***************/
+	reinitializeAppResponse(client, req);
+	sendHeaderToApp(client, req);
+}
+
 void
 Controller::initiateSession(Client *client, Request *req) {
 	TRACE_POINT();
 	req->sessionCheckoutTry++;
 	try {
-		req->session->initiate(false);
+		if (!req->session->initiate()) {
+			ev_io connectedWatcher;
```

Comment:
You can't allocate this on the stack. It has to stay alive while waiting is in progress. Should be in the Request struct.